### PR TITLE
[dynamo] fix polyfill for user defined constructor `__new__`

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133822

In `cls->tp_call`, if `cls->tp_new` does not return an instance of class `cls`, then `cls->tp_init` is not called on the new instance.

Related PR:

- #132977

cc @rec @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @xmfan

### CODE REVIEW

File: torch/_dynamo/decorators.py [hunk 1]
```diff
@@ -159,6 +165,166 @@ def forbid_in_graph(fn):
     return fn
 
 
+def substitute_in_graph(
```
Comment (jansel) [inline torch/_dynamo/decorators.py:168]: Wasn't this already added in #133712

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/decorators.py (hunk 1 above)
Comment (XuehaiPan) [reply torch/_dynamo/decorators.py:168]: I think this is a bug for `ghstack`. I'll try to figure it out.

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -523,7 +523,7 @@ def is_standard_new(self):
         new_fn = inspect.getattr_static(self.value, "__new__", None)
         if isinstance(new_fn, staticmethod):
             new_fn = new_fn.__func__
-        return new_fn in (object.__new__, Generic.__new__)
+        return new_fn in {object.__new__, Generic.__new__}
```
Comment (jansel) [inline torch/_dynamo/variables/user_defined.py:526]: Why this change?  Using a set like this will only be faster if you move it into global scope, since you still need to construct the set on every call to the function (which is O(n), but also involves a bunch of hashing so is higher slower than tuple creation).

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/user_defined.py (hunk 2 above)
Comment (XuehaiPan) [reply torch/_dynamo/variables/user_defined.py:526]: > Using a set like this will only be faster if you move it into global scope

That makes sense. Reverted.

**Review by jansel — APPROVED** [review]
